python/ray/dag/compiled_dag_node.py

```diff
@@ -2443,7 +2443,14 @@ def teardown(self, kill_actors: bool = False):
 
         monitor = getattr(self, "_monitor", None)
         if monitor is not None:
+            from ray.dag import DAGContext
+
+            ctx = DAGContext.get_current()
             monitor.teardown(kill_actors=kill_actors)
+            monitor.join(timeout=ctx.teardown_timeout)
```

Comment:
Based on the PR description, 

> To prevent this from happening, this fix joins the monitor thread before destructing the CoreWorker.

I don't understand when the `CoreWorker` destruction happens and why calling `monitor.join` in `__del__` ensures that the monitor thread joins before the `CoreWorker` is destructed.

I guess the CoreWorker will be terminated when `ray.shutdown()` is called.

Reply:
Yes, CoreWorker is destructed in `ray.shutdown()`, and the monitor thread could still be running at that point.